File: src/chinese_char_counter_mcp/counter.py
"""中文字数统计的核心逻辑。

设计约束：
- 纯函数，不读写文件、不打印任何内容（MCP STDIO 传输里 stdout 是协议通道，
  任何 print 都会搅坏 JSON-RPC 帧，所以本模块与上层服务都不使用 print）。
- 不依赖 mcp 包，可被其他项目直接 import 复用，也便于纯单元测试。
"""

from __future__ import annotations

import unicodedata
from bisect import bisect_right
from typing import Any, Dict, Sequence

# 计入“中文”的 Unicode 码位区间：CJK 统一表意文字、各扩展区、兼容表意文字，
# 以及表意数字零“〇”。区间按起始码位升序排列，供 bisect 二分查找。
HAN_RANGES: Sequence[tuple[int, int]] = (
    (0x3007, 0x3007),    # 〇 表意数字零
    (0x3400, 0x4DBF),    # 扩展 A
    (0x4E00, 0x9FFF),    # 基本区
    (0xF900, 0xFAFF),    # 兼容表意文字
    (0x20000, 0x2A6DF),  # 扩展 B
    (0x2A700, 0x2B73F),  # 扩展 C
    (0x2B740, 0x2B81F),  # 扩展 D
    (0x2B820, 0x2CEAF),  # 扩展 E
    (0x2CEB0, 0x2EBEF),  # 扩展 F
    (0x2EBF0, 0x2EE5F),  # 扩展 I
    (0x2F800, 0x2FA1F),  # 兼容表意文字补充
    (0x30000, 0x3134F),  # 扩展 G
    (0x31350, 0x323AF),  # 扩展 H
)

_HAN_STARTS = tuple(start for start, _ in HAN_RANGES)
_HAN_ENDS = tuple(end for _, end in HAN_RANGES)

# 单次调用的文本长度上限与批量条数上限，避免客户端一次丢进超大 payload。
MAX_TEXT_LENGTH = 200_000
MAX_BATCH_SIZE = 500

# Unicode 类别归类的目标桶（顺序即返回结果中的字段顺序）。
CATEGORIES = ("chinese", "letters", "digits", "punctuation", "spaces", "other")

# 这些控制字符本质是空白，归入 spaces 桶。
_ASCII_WHITESPACE = "\t\n\r\v\f"


def is_chinese_char(char: str) -> bool:
    """判断单个字符是否计入“中文”。

    计入：CJK 统一表意文字（基本区、扩展 A~I、兼容表意文字）与“〇”。
    不计入：标点、空格、英文字母、阿拉伯数字、emoji 及其他文字（假名、谚文等）。
    """
    if not char:
        return False
    codepoint = ord(char)
    index = bisect_right(_HAN_STARTS, codepoint) - 1
    return index >= 0 and codepoint <= _HAN_ENDS[index]


def classify_char(char: str) -> str:
    """把字符归入 CATEGORIES 中的一个桶。"""
    if is_chinese_char(char):
        return "chinese"
    if char in _ASCII_WHITESPACE:
        return "spaces"
    category = unicodedata.category(char)
    first = category[0]
    if first == "L":
        return "letters"
    if first == "N":
        return "digits"
    if first == "Z":
        return "spaces"
    if first == "P":
        return "punctuation"
    return "other"


def count_text(text: str) -> Dict[str, Any]:
    """统计文本的中文字数并给出分项明细。

    :raises TypeError: text 不是字符串
    :raises ValueError: text 长度超过 MAX_TEXT_LENGTH
    """
    if not isinstance(text, str):
        raise TypeError("text must be a string")
    if len(text) > MAX_TEXT_LENGTH:
        raise ValueError(
            "text is too long: %d characters (limit %d)" % (len(text), MAX_TEXT_LENGTH)
        )

    breakdown = {name: 0 for name in CATEGORIES}
    for char in text:
        breakdown[classify_char(char)] += 1

    chinese_count = breakdown["chinese"]
    total = len(text)
    return {
        "chinese_count": chinese_count,
        "total_characters": total,
        "chinese_ratio": round(chinese_count / total, 4) if total else 0.0,
        "breakdown": breakdown,
    }


def extract_chinese(text: str) -> str:
    """抽取文本中所有计入“中文”的字符，按原顺序拼接返回。"""
    if not isinstance(text, str):
        raise TypeError("text must be a string")
    if len(text) > MAX_TEXT_LENGTH:
        raise ValueError(
            "text is too long: %d characters (limit %d)" % (len(text), MAX_TEXT_LENGTH)
        )
    return "".join(char for char in text if is_chinese_char(char))


def count_batch(texts: Sequence[str]) -> Dict[str, Any]:
    """批量统计多条文本的中文字数。

    :raises TypeError: texts 不是列表/元组，或其中某项不是字符串
    :raises ValueError: 条数超过 MAX_BATCH_SIZE，或某条文本超长
    """
    if not isinstance(texts, (list, tuple)):
        raise TypeError("texts must be a list of strings")
    if len(texts) > MAX_BATCH_SIZE:
        raise ValueError(
            "too many texts: %d (limit %d)" % (len(texts), MAX_BATCH_SIZE)
        )

    results = []
    total_chinese = 0
    for index, item in enumerate(texts):
        item_counts = count_text(item)
        total_chinese += item_counts["chinese_count"]
        results.append(
            {
                "index": index,
                "total_characters": item_counts["total_characters"],
                "chinese_count": item_counts["chinese_count"],
                "chinese_ratio": item_counts["chinese_ratio"],
            }
        )
    return {
        "results": results,
        "text_count": len(results),
        "total_chinese_count": total_chinese,
    }

File: npm/lib/counter.js
/**
 * 中文字数统计核心逻辑（与仓库里的 Python 版逐字对齐）。
 *
 * 纯函数，不打印任何内容 —— MCP STDIO 传输里 stdout 是协议通道，
 * 任何 console.log 都会搅坏 JSON-RPC 帧（日志一律走 stderr）。
 */

// 计入「中文」的码位区间：CJK 统一表意文字、各扩展区、兼容表意文字，以及表意数字零「〇」。
// 与 Python 版 HAN_RANGES 完全一致，升序排列。
export const HAN_RANGES = [
  [0x3007, 0x3007],    // 〇 表意数字零
  [0x3400, 0x4dbf],    // 扩展 A
  [0x4e00, 0x9fff],    // 基本区
  [0xf900, 0xfaff],    // 兼容表意文字
  [0x20000, 0x2a6df],  // 扩展 B
  [0x2a700, 0x2b73f],  // 扩展 C
  [0x2b740, 0x2b81f],  // 扩展 D
  [0x2b820, 0x2ceaf],  // 扩展 E
  [0x2ceb0, 0x2ebef],  // 扩展 F
  [0x2ebf0, 0x2ee5f],  // 扩展 I
  [0x2f800, 0x2fa1f],  // 兼容表意文字补充
  [0x30000, 0x3134f],  // 扩展 G
  [0x31350, 0x323af],  // 扩展 H
];

export const MAX_TEXT_LENGTH = 200000;
export const MAX_BATCH_SIZE = 500;
export const CATEGORIES = ["chinese", "letters", "digits", "punctuation", "spaces", "other"];

const ASCII_WHITESPACE = new Set(["\t", "\n", "\r", "\v", "\f"]);

export function isChineseChar(char) {
  if (!char) return false;
  const cp = char.codePointAt(0);
  for (const [lo, hi] of HAN_RANGES) {
    if (cp >= lo && cp <= hi) return true;
    if (cp < lo) return false; // 区间升序，提前退出
  }
  return false;
}

export function classifyChar(char) {
  if (isChineseChar(char)) return "chinese";
  if (ASCII_WHITESPACE.has(char)) return "spaces";
  // 顺序与 Python 版一致：L → N → Z → P
  if (/\p{L}/u.test(char)) return "letters";
  if (/\p{N}/u.test(char)) return "digits";
  if (/\p{Z}/u.test(char)) return "spaces";
  if (/\p{P}/u.test(char)) return "punctuation";
  return "other";
}

/** 复刻 Python round(x, 4) 的「四舍六入五成双」行为。 */
export function roundRatio(value, digits = 4) {
  const factor = 10 ** digits;
  const scaled = value * factor;
  const floor = Math.floor(scaled);
  const diff = scaled - floor;
  let rounded;
  if (Math.abs(diff - 0.5) < 1e-9) {
    rounded = floor % 2 === 0 ? floor : floor + 1;
  } else {
    rounded = Math.round(scaled);
  }
  return rounded / factor;
}

function emptyBreakdown() {
  return { chinese: 0, letters: 0, digits: 0, punctuation: 0, spaces: 0, other: 0 };
}

export function countText(text) {
  if (typeof text !== "string") throw new TypeError("text must be a string");
  // 按「码点」而不是 UTF-16 码元计数：Python 的 len(str) 数的是码点，
  // 用 text.length 会让 emoji / 扩展 B 生僻字（代理对）多算一倍。
  const characters = [...text];
  if (characters.length > MAX_TEXT_LENGTH) {
    throw new RangeError(
      `text is too long: ${characters.length} characters (limit ${MAX_TEXT_LENGTH})`,
    );
  }
  const breakdown = emptyBreakdown();
  for (const char of characters) breakdown[classifyChar(char)] += 1;
  const chineseCount = breakdown.chinese;
  const total = characters.length;
  return {
    chinese_count: chineseCount,
    total_characters: total,
    chinese_ratio: total ? roundRatio(chineseCount / total) : 0.0,
    breakdown,
  };
}

export function extractChinese(text) {
  if (typeof text !== "string") throw new TypeError("text must be a string");
  const characters = [...text];
  if (characters.length > MAX_TEXT_LENGTH) {
    throw new RangeError(
      `text is too long: ${characters.length} characters (limit ${MAX_TEXT_LENGTH})`,
    );
  }
  let out = "";
  for (const char of characters) if (isChineseChar(char)) out += char;
  return out;
}

export function countBatch(texts) {
  if (!Array.isArray(texts)) throw new TypeError("texts must be a list of strings");
  if (texts.length > MAX_BATCH_SIZE) {
    throw new RangeError(`too many texts: ${texts.length} (limit ${MAX_BATCH_SIZE})`);
  }
  const results = [];
  let totalChinese = 0;
  texts.forEach((item, index) => {
    const counts = countText(item);
    totalChinese += counts.chinese_count;
    results.push({
      index,
      total_characters: counts.total_characters,
      chinese_count: counts.chinese_count,
      chinese_ratio: counts.chinese_ratio,
    });
  });
  return { results, text_count: results.length, total_chinese_count: totalChinese };
}
